File: core/src/main/scala/kafka/log/streamaspect/ElasticProducerStateManager.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kafka.log.streamaspect

import kafka.log._
import org.apache.kafka.common.TopicPartition
import org.apache.kafka.common.protocol.types.SchemaException
import org.apache.kafka.common.utils.Time

import java.io.File
import java.nio.ByteBuffer
import java.util
import java.util.concurrent.{CompletableFuture, ConcurrentSkipListMap}
import scala.collection.mutable
import scala.jdk.CollectionConverters.CollectionHasAsScala

/**
 * ElasticProducerStateManager. Temporarily, we only persist the last snapshot.
 *
 * @param snapshotsMap All valid snapshots.
 */
class ElasticProducerStateManager(
                                   override val topicPartition: TopicPartition,
                                   __logDir: File,
                                   override val maxTransactionTimeoutMs: Int,
                                   override val producerStateManagerConfig: ProducerStateManagerConfig,
                                   override val time: Time,
                                   val snapshotsMap: util.NavigableMap[java.lang.Long, ByteBuffer],
                                   val persistFun: MetaKeyValue => CompletableFuture[Void]
                                 ) extends ProducerStateManager(topicPartition, __logDir, maxTransactionTimeoutMs, producerStateManagerConfig, time) {

  this.logIdent = s"[ElasticProducerStateManager partition=$topicPartition] "

  override protected def loadSnapshots(): ConcurrentSkipListMap[java.lang.Long, SnapshotFile] = {
    val tm = new ConcurrentSkipListMap[java.lang.Long, SnapshotFile]()
    snapshotsMap.forEach { case (offset, meta) =>
      tm.put(offset, SnapshotFile(new File(s"$offset.snapshot")))
    }
    tm
  }

  override protected def loadFromSnapshot(logStartOffset: Long, currentTime: Long): Unit = {
    while (true) {
      latestSnapshotFile match {
        case Some(snapshot) =>
          try {
            info(s"Loading producer state from snapshot file '$snapshot'")
            val loadedProducers = readSnapshot(snapshot.file).filter { producerEntry => !isProducerExpired(currentTime, producerEntry) }
            loadedProducers.foreach(loadProducerEntry)
            lastSnapOffset = snapshot.offset
            lastMapOffset = lastSnapOffset
            updateOldestTxnTimestamp()
            return
          } catch {
            case e: CorruptSnapshotException =>
              warn(s"Failed to load producer snapshot from '${snapshot.file}': ${e.getMessage}")
              removeAndDeleteSnapshot(snapshot.offset)
          }
        case None =>
          lastSnapOffset = logStartOffset
          lastMapOffset = logStartOffset
          return
      }
    }
  }


  override def truncateAndReload(logStartOffset: Long, logEndOffset: Long, currentTimeMs: Long): Unit = {
    // remove all out of range snapshots
    snapshots.values().asScala.foreach { snapshot =>
      if (snapshot.offset > logEndOffset || snapshot.offset <= logStartOffset) {
        removeAndDeleteSnapshot(snapshot.offset)
      }
    }

    if (logEndOffset != mapEndOffset) {
      producers.clear()
      ongoingTxns.clear()
      updateOldestTxnTimestamp()

      // since we assume that the offset is less than or equal to the high watermark, it is
      // safe to clear the unreplicated transactions
      unreplicatedTxns.clear()
    }
    loadFromSnapshot(logStartOffset, currentTimeMs)
  }

  override def takeSnapshot(): Unit = {
    takeSnapshot0()
  }

  def takeSnapshot0(): CompletableFuture[Void] = {
    // If not a new offset, then it is not worth taking another snapshot
    if (lastMapOffset > lastSnapOffset) {
      val snapshotFile = SnapshotFile(UnifiedLog.producerSnapshotFile(_logDir, lastMapOffset))
      val start = time.hiResClockMs()
      val rst = writeSnapshot(snapshotFile.offset, producers)
      snapshots.put(snapshotFile.offset, snapshotFile)
      info(s"Wrote producer snapshot at offset $lastMapOffset with ${producers.size} producer ids in ${time.hiResClockMs() - start} ms.")


      // Update the last snap offset according to the serialized map
      lastSnapOffset = lastMapOffset
      rst
    } else {
      CompletableFuture.completedFuture(null)
    }
  }

  /**
   * Only keep the last snapshot which offset less than or equals to the recoveryPointCheckpoint
   */
  def takeSnapshotAndRemoveExpired(recoveryPointCheckpoint: Long): CompletableFuture[Void] = {
    val lastSnapshotOffset = snapshotsMap.floorKey(recoveryPointCheckpoint)
    if (lastSnapshotOffset != null) {
      val expiredSnapshotOffsets = new util.ArrayList[java.lang.Long](snapshotsMap.headMap(lastSnapshotOffset, false).keySet())
      expiredSnapshotOffsets.forEach(offset => {
        snapshotsMap.remove(offset)
        snapshots.remove(offset)
      })
    }
    takeSnapshot0()
  }

  private def writeSnapshot(offset: Long, entries: mutable.Map[Long, ProducerStateEntry]): CompletableFuture[Void] = {
    val buffer = ProducerStateManager.writeSnapshotToBuffer(entries)
    val rawSnapshot: Array[Byte] = new Array[Byte](buffer.remaining())
    buffer.get(rawSnapshot)

    snapshotsMap.put(offset, ByteBuffer.wrap(rawSnapshot))
    val meta = new ElasticPartitionProducerSnapshotsMeta(snapshotsMap)
    persistFun(MetaKeyValue.of(MetaStream.PRODUCER_SNAPSHOTS_META_KEY, meta.encode()))
  }

  private def readSnapshot(file: File): Iterable[ProducerStateEntry] = {
    val offset = LocalLog.offsetFromFile(file)
    if (!snapshotsMap.containsKey(offset)) {
      throw new CorruptSnapshotException(s"Snapshot not found")
    }

    try {
      ProducerStateManager.readSnapshotFromBuffer(snapshotsMap.get(offset).array())
    } catch {
      case e: SchemaException =>
        throw new CorruptSnapshotException(s"Snapshot failed schema validation: ${e.getMessage}")
    }
  }

  // do nothing
  override def updateParentDir(parentDir: File): Unit = {}

  override protected def removeAndDeleteSnapshot(snapshotOffset: Long): Unit = {
    deleteSnapshot(snapshotOffset)
  }

  override private[log] def removeAndMarkSnapshotForDeletion(snapshotOffset: Long): Option[SnapshotFile] = {
    deleteSnapshot(snapshotOffset)
    None
  }

  override def deleteSnapshotFile(snapshotFile: SnapshotFile): Unit = {
    deleteSnapshot(snapshotFile.offset)
  }

  private def deleteSnapshot(snapshotOffset: Long): Unit = {
    // the real deletion will happens after meta compaction.
    snapshots.remove(snapshotOffset)
    val deleted = snapshotsMap.remove(snapshotOffset)
    if (deleted != null && isDebugEnabled) {
      debug(s"Deleted producer snapshot file '$snapshotOffset' for partition $topicPartition")
    }
  }
}

object ElasticProducerStateManager {
  def apply(
             topicPartition: TopicPartition,
             logDir: File,
             maxTransactionTimeoutMs: Int,
             producerStateManagerConfig: ProducerStateManagerConfig,
             time: Time,
             snapshotMap: util.Map[java.lang.Long, ByteBuffer],
             persistFun: MetaKeyValue => CompletableFuture[Void]
           ): ElasticProducerStateManager = {
    val stateManager = new ElasticProducerStateManager(
      topicPartition,
      logDir,
      maxTransactionTimeoutMs,
      producerStateManagerConfig,
      time,
      new util.TreeMap[java.lang.Long, ByteBuffer](snapshotMap),
      persistFun
    )
    stateManager
  }
}

File: core/src/main/scala/kafka/log/streamaspect/ElasticProducerStateManager.java
/*
 * Copyright 2025, AutoMQ HK Limited.
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kafka.log.streamaspect;

import org.apache.kafka.common.TopicPartition;
import org.apache.kafka.common.errors.OutOfOrderSequenceException;
import org.apache.kafka.common.utils.Time;
import org.apache.kafka.storage.internals.log.AppendOrigin;
import org.apache.kafka.storage.internals.log.CorruptSnapshotException;
import org.apache.kafka.storage.internals.log.LogFileUtils;
import org.apache.kafka.storage.internals.log.ProducerAppendInfo;
import org.apache.kafka.storage.internals.log.ProducerStateEntry;
import org.apache.kafka.storage.internals.log.ProducerStateManager;
import org.apache.kafka.storage.internals.log.ProducerStateManagerConfig;
import org.apache.kafka.storage.internals.log.SnapshotFile;
import org.apache.kafka.storage.internals.log.VerificationStateEntry;

import java.io.File;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.NavigableMap;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentSkipListMap;

public class ElasticProducerStateManager extends ProducerStateManager {
    public static final long AWAIT_SEQ_ZERO_TIMEOUT = 120000L;
    private final PersistSnapshots persistSnapshots;
    private final long createTimestamp;

    public ElasticProducerStateManager(
        TopicPartition topicPartition,
        File logDir,
        int maxTransactionTimeoutMs,
        ProducerStateManagerConfig producerStateManagerConfig,
        Time time,

        NavigableMap<Long, ByteBuffer> snapshotsMap,
        PersistSnapshots persistSnapshots
    ) throws IOException {
        super(topicPartition, logDir, maxTransactionTimeoutMs, producerStateManagerConfig, time);
        this.snapshotsMap = snapshotsMap;
        this.persistSnapshots = persistSnapshots;
        // ProducerStateManager will loadSnapshots in constructor, but at that moment, snapshotsMap is null.
        // So we need to call loadSnapshots again to load snapshots from snapshotsMap
        this.snapshots = loadSnapshots();
        this.createTimestamp = time.milliseconds();
    }

    @Override
    protected ConcurrentSkipListMap<Long, SnapshotFile> loadSnapshots() {
        NavigableMap<Long, ByteBuffer> snapshotsMap = this.snapshotsMap;
        if (snapshotsMap.isEmpty()) {
            // ProducerStateManager will loadSnapshots in constructor
            return new ConcurrentSkipListMap<>();
        }
        ConcurrentSkipListMap<Long, SnapshotFile> offsetToSnapshots = new ConcurrentSkipListMap<>();
        snapshotsMap.forEach((offset, snapshot) -> {
            File snapshotFile = new File(offset + ".snapshot");
            offsetToSnapshots.put(offset, new MockSnapshotFile(snapshotFile, offset));
        });
        return offsetToSnapshots;
    }

    @Override
    public void removeStraySnapshots(Collection<Long> segmentBaseOffsets) throws IOException {
        // it's ok to call super method
        super.removeStraySnapshots(segmentBaseOffsets);
    }

    @Override
    public void truncateAndReload(long logStartOffset, long logEndOffset, long currentTimeMs) throws IOException {
        // it's ok to call super method
        super.truncateAndReload(logStartOffset, logEndOffset, currentTimeMs);
        // TODO: check whether need to call onLogStartOffsetIncremented
    }

    @Override
    public ProducerAppendInfo prepareUpdate(long producerId, AppendOrigin origin) {
        ProducerStateEntry currentEntry = lastEntry(producerId).orElse(ProducerStateEntry.empty(producerId));
        return new ProducerAppendInfoExt(topicPartition, producerId, currentEntry, origin, verificationStateEntry(producerId));    }

    @Override
    protected List<ProducerStateEntry> readSnapshot(File file) throws IOException {
        long offset = LogFileUtils.offsetFromFile(file);
        if (!snapshotsMap.containsKey(offset)) {
            throw new CorruptSnapshotException("Snapshot" + offset + " not found");
        }
        return ProducerStateManager.readSnapshot0(file, snapshotsMap.get(offset).array());
    }

    @Override
    public void updateParentDir(File parentDir) {
        // noop implementation
    }

    @Override
    protected void removeAndDeleteSnapshot(long snapshotOffset) {
        deleteSnapshot(snapshotOffset);
    }

    @Override
    public Optional<SnapshotFile> removeAndMarkSnapshotForDeletion(long snapshotOffset) {
        return deleteSnapshot(snapshotOffset);
    }

    @Override
    public void takeSnapshot() throws IOException {
        try {
            takeSnapshot0().get();
        } catch (Throwable e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public Optional<File> takeSnapshot(boolean sync) throws IOException {
        try {
            Optional<File> rst;
            if (lastMapOffset > lastSnapOffset) {
                rst = Optional.of(LogFileUtils.producerSnapshotFile(logDir, lastMapOffset));
            } else {
                rst = Optional.empty();
            }
            CompletableFuture<Void> cf = takeSnapshot0();
            if (sync) {
                cf.get();
            }
            return rst;
        } catch (Throwable e) {
            throw new RuntimeException(e);
        }
    }

    CompletableFuture<Void> takeSnapshot0() {
        // If not a new offset, then it is not worth taking another snapshot
        if (lastMapOffset > lastSnapOffset) {
            SnapshotFile snapshotFile = new MockSnapshotFile(LogFileUtils.producerSnapshotFile(logDir, lastMapOffset), lastMapOffset);
            CompletableFuture<Void> rst = writeSnapshot(snapshotFile.offset, producers);
            snapshots.put(snapshotFile.offset, snapshotFile);
            // Update the last snap offset according to the serialized map
            lastSnapOffset = lastMapOffset;
            return rst;
        } else {
            return CompletableFuture.completedFuture(null);
        }
    }


    /**
     * Only keep the last snapshot which offset less than or equals to the recoveryPointCheckpoint
     */
    public CompletableFuture<Void> takeSnapshotAndRemoveExpired(long recoveryPointCheckpoint) {
        Long lastSnapshotOffset = snapshotsMap.floorKey(recoveryPointCheckpoint);
        if (lastSnapshotOffset != null) {
            List<Long> expiredSnapshotOffsets = new ArrayList<>(snapshotsMap.headMap(lastSnapshotOffset, false).keySet());
            expiredSnapshotOffsets.forEach(offset -> {
                snapshotsMap.remove(offset);
                snapshots.remove(offset);
            });
        }
        return takeSnapshot0();
    }

    private CompletableFuture<Void> writeSnapshot(long offset, Map<Long, ProducerStateEntry> entries) {
        try {
            ByteBuffer buffer = ProducerStateManager.writeSnapshot(null, entries, false);
            byte[] rawSnapshot = new byte[buffer.remaining()];
            buffer.get(rawSnapshot);

            snapshotsMap.put(offset, ByteBuffer.wrap(rawSnapshot));
            ElasticPartitionProducerSnapshotsMeta meta = new ElasticPartitionProducerSnapshotsMeta(snapshotsMap);
            return persistSnapshots.persist(MetaKeyValue.of(MetaStream.PRODUCER_SNAPSHOTS_META_KEY, meta.encode()));
        } catch (IOException e) {
            return CompletableFuture.failedFuture(e);
        }
    }


    private Optional<SnapshotFile> deleteSnapshot(long snapshotOffset) {
        SnapshotFile snapshotFile = snapshots.remove(snapshotOffset);
        snapshotsMap.remove(snapshotOffset);
        return Optional.ofNullable(snapshotFile);
    }

    public interface PersistSnapshots {
        CompletableFuture<Void> persist(MetaKeyValue kv);
    }

    NavigableMap<Long, ByteBuffer> snapshotsMap() {
        return snapshotsMap;
    }

    class MockSnapshotFile extends SnapshotFile {

        public MockSnapshotFile(File file, long offset) {
            super(file, offset);
        }

        @Override
        public boolean deleteIfExists() throws IOException {
            snapshotsMap.remove(offset);
            return true;
        }

        @Override
        public void renameToDelete() {
            snapshotsMap.remove(offset);
        }

        @Override
        public void updateParentDir(File parentDir) {
            // noop implementation
        }
    }

    class ProducerAppendInfoExt extends ProducerAppendInfo {
        public ProducerAppendInfoExt(TopicPartition topicPartition, long producerId, ProducerStateEntry currentEntry,
            AppendOrigin origin, VerificationStateEntry verificationStateEntry) {
            super(topicPartition, producerId, currentEntry, origin, verificationStateEntry);
        }

        @Override
        protected void checkSequence(short producerEpoch, int appendFirstSeq, long offset) {
            if (currentEntry.isEmpty() && updatedEntry.isEmpty() && appendFirstSeq != 0
                // await sequence 0 message append timeout and retry
                && time.milliseconds() - createTimestamp < AWAIT_SEQ_ZERO_TIMEOUT
            ) {
                throw new OutOfOrderSequenceException("Invalid sequence number for new created log, producer " + producerId() + " " +
                    "at offset " + offset + " in partition " + topicPartition + ": " + producerEpoch + " (request epoch), " + appendFirstSeq + " (seq. number), " +
                    updatedEntry.producerEpoch() + " (current producer epoch)");
            }
            super.checkSequence(producerEpoch, appendFirstSeq, offset);
        }
    }
}
